File: ios/LibVlcPlayerView.swift
import ExpoModulesCore
#if os(tvOS)
    import TVVLCKit
#else
    import MobileVLCKit
#endif
import UIKit

let defaultPlayerRate: Float = 1.0
let defaultPlayerTime: Int = 0
let defaultPlayerScale: Float = 0.0

let minPlayerVolume: Int = 0
let maxPlayerVolume: Int = 100
let playerVolumeStep: Int = 10

let dialogCustomUI: Bool = true

enum ResizeMode: String {
    case contain
    case cover
    case fill
}

class LibVlcPlayerView: ExpoView {
    private let playerView = UIView()

    var mediaPlayer: VLCMediaPlayer?
    var vlcDialog: VLCDialogProvider?
    var vlcDialogRef: NSValue?

    var mediaLength: Int32 = 0
    private var oldVolume: Int = maxPlayerVolume

    private var shouldCreate: Bool = false
    var firstPlay: Bool = false
    var firstTime: Bool = false

    let onBuffering = EventDispatcher()
    let onPlaying = EventDispatcher()
    let onPaused = EventDispatcher()
    let onStopped = EventDispatcher()
    let onEndReached = EventDispatcher()
    let onEncounteredError = EventDispatcher()
    let onTimeChanged = EventDispatcher()
    let onPositionChanged = EventDispatcher()
    let onESAdded = EventDispatcher()
    let onDialogDisplay = EventDispatcher()
    let onFirstPlay = EventDispatcher()
    let onBackground = EventDispatcher()

    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)

        MediaPlayerManager.shared.registerPlayerView(self)

        clipsToBounds = true
        playerView.backgroundColor = .black
        addSubview(playerView)
    }

    deinit {
        MediaPlayerManager.shared.unregisterPlayerView(self)
        destroyPlayer()
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        playerView.frame = bounds

        // Reapply resize mode when view size changes (e.g., rotation)
        if mediaPlayer != nil {
            applyResizeMode()
        }
    }

    func createPlayer() {
        if !shouldCreate {
            return
        }

        destroyPlayer()

        guard let source = source else { return }

        if autoplay {
            options.removeStartPausedOption()
        }

        mediaPlayer = VLCMediaPlayer(options: options)
        mediaPlayer!.drawable = playerView
        mediaPlayer!.delegate = self

        let library = mediaPlayer!.libraryInstance
        vlcDialog = VLCDialogProvider(library: library, customUI: dialogCustomUI)
        vlcDialog!.customRenderer = self

        guard let url = URL(string: source) else {
            let error = ["error": "Invalid source, media could not be set"]
            onEncounteredError(error)
            return
        }

        mediaPlayer!.media = VLCMedia(url: url)

        addPlayerSlaves()

        mediaPlayer!.play()

        shouldCreate = false
        firstPlay = true
        firstTime = true
    }

    func destroyPlayer() {
        vlcDialogRef = nil
        vlcDialog = nil
        mediaPlayer?.media = nil
        mediaPlayer?.delegate = nil
        mediaPlayer?.drawable = nil
        mediaPlayer = nil
    }

    func setPlayerTracks() {
        if let player = mediaPlayer {
            let audioTrackIndex = tracks?.audio ?? Int(player.currentAudioTrackIndex)
            let videoTrackIndex = tracks?.video ?? Int(player.currentVideoTrackIndex)
            let videoSubTitleIndex = tracks?.subtitle ?? Int(player.currentVideoSubTitleIndex)

            player.currentAudioTrackIndex = Int32(audioTrackIndex)
            player.currentVideoTrackIndex = Int32(videoTrackIndex)
            player.currentVideoSubTitleIndex = Int32(videoSubTitleIndex)
        }
    }

    func addPlayerSlaves() {
        for slave in slaves {
            let source = slave.source
            let type = slave.type
            let slaveType = type == "subtitle" ?
                VLCMediaPlaybackSlaveType.subtitle :
                VLCMediaPlaybackSlaveType.audio
            let selected = slave.selected ?? false

            guard let url = URL(string: source) else {
                let error = ["error": "Invalid slave, \(type) could not be added"]
                onEncounteredError(error)
                continue
            }

            mediaPlayer?.addPlaybackSlave(url, type: slaveType, enforce: selected)
        }
    }

    func getMediaTracks() -> MediaTracks {
        var mediaTracks = MediaTracks()

        if let player = mediaPlayer {
            var audioTracks: [Track] = []

            if let audios = player.audioTrackNames as? [String] {
                if let audioIndexes = player.audioTrackIndexes as? [NSNumber] {
                    for (index, trackName) in audios.enumerated() {
                        let trackId = audioIndexes[index].intValue
                        let track = Track(id: trackId, name: trackName)
                        audioTracks.append(track)
                    }
                }
            }

            var videoTracks: [Track] = []

            if let videos = player.videoTrackNames as? [String] {
                if let videoIndexes = player.videoTrackIndexes as? [NSNumber] {
                    for (index, trackName) in videos.enumerated() {
                        let trackId = videoIndexes[index].intValue
                        let track = Track(id: trackId, name: trackName)
                        videoTracks.append(track)
                    }
                }
            }

            var subtitleTracks: [Track] = []

            if let subtitles = player.videoSubTitlesNames as? [String] {
                if let subtitleIndexes = player.videoSubTitlesIndexes as? [NSNumber] {
                    for (index, trackName) in subtitles.enumerated() {
                        let trackId = subtitleIndexes[index].intValue
                        let track = Track(id: trackId, name: trackName)
                        subtitleTracks.append(track)
                    }
                }
            }

            mediaTracks = MediaTracks(
                audio: audioTracks,
                video: videoTracks,
                subtitle: subtitleTracks
            )
        }

        return mediaTracks
    }

    func getMediaInfo() -> MediaInfo {
        var mediaInfo = MediaInfo()

        if let player = mediaPlayer {
            let video = player.videoSize
            let length = player.media?.length.intValue ?? 0
            let seekable = player.isSeekable
            let mediaTracks = getMediaTracks()

            mediaInfo = MediaInfo(
                width: Int(video.width),
                height: Int(video.height),
                length: Double(length),
                seekable: seekable,
                tracks: mediaTracks
            )

            mediaLength = length
        }

        return mediaInfo
    }

    func setupPlayer() {
        if let player = mediaPlayer {
            setPlayerTracks()

            if volume != maxPlayerVolume || mute {
                let newVolume = mute ?
                    minPlayerVolume :
                    volume

                player.audio?.volume = Int32(newVolume)
            }

            if rate != defaultPlayerRate {
                player.rate = rate
            }

            if time != defaultPlayerTime {
                player.time = VLCTime(int: Int32(time))
            }

            // Apply scale and aspectRatio only if contentFit is contain (default)
            // Otherwise, applyResizeMode() will manage these settings
            if contentFit == .contain {
                if scale != defaultPlayerScale {
                    player.scaleFactor = scale
                }

                if let aspectRatio = aspectRatio {
                    aspectRatio.withCString { cString in
                        player.videoAspectRatio = UnsafeMutablePointer(mutating: cString)
                    }
                }
            }

            applyResizeMode()

            time = defaultPlayerTime
        }
    }

    var source: String? {
        didSet {
            shouldCreate = source != oldValue
        }
    }

    var options: [String] = .init() {
        didSet {
            shouldCreate = options != oldValue
        }
    }

    var tracks: Tracks? {
        didSet {
            setPlayerTracks()
        }
    }

    private var _slaves: [Slave] = .init()

    var slaves: [Slave] {
        get { _slaves }
        set {
            let newSlaves = newValue.filter { slave in !_slaves.contains(slave) }

            _slaves += newSlaves

            if !newSlaves.isEmpty {
                addPlayerSlaves()
            }
        }
    }

    var scale: Float = defaultPlayerScale {
        didSet {
            // Only apply scale directly if contentFit is contain (default)
            // Otherwise, contentFit manages scaleFactor
            if contentFit == .contain {
                mediaPlayer?.scaleFactor = scale
            }
        }
    }

    var aspectRatio: String? {
        didSet {
            // Only apply aspectRatio directly if contentFit is contain (default)
            // Otherwise, contentFit manages videoAspectRatio
            if contentFit == .contain {
                if let aspectRatio = aspectRatio {
                    aspectRatio.withCString { cString in
                        mediaPlayer?.videoAspectRatio = UnsafeMutablePointer(mutating: cString)
                    }
                } else {
                    mediaPlayer?.videoAspectRatio = nil
                }
            }
        }
    }

    var contentFit: ResizeMode = .contain {
        didSet {
            applyResizeMode()
        }
    }

    func applyResizeMode() {
        guard let player = mediaPlayer else { return }

        // Clear previous settings first to avoid conflicts
        player.videoAspectRatio = nil
        player.videoCropGeometry = nil
        player.scaleFactor = 0.0

        // Guard against invalid bounds - will be reapplied in layoutSubviews when bounds are set
        guard bounds.width > 0, bounds.height > 0 else { return }

        switch contentFit {
        case .contain:
            // Already cleared above, nothing more needed
            break

        case .cover:
            // Aspect fill - video fills view, may crop edges
            let videoSize = player.videoSize
            guard videoSize.width > 0, videoSize.height > 0 else {
                // Video not ready yet, will be applied on esAdded or in setupPlayer()
                return
            }

            let viewAspect = bounds.width / bounds.height
            let videoAspect = videoSize.width / videoSize.height

            if videoAspect > viewAspect {
                // Video is wider - crop width
                let cropWidth = Int(videoSize.height * viewAspect)
                let cropGeometry = "\(cropWidth):\(Int(videoSize.height))"
                cropGeometry.withCString { cString in
                    player.videoCropGeometry = UnsafeMutablePointer(mutating: cString)
                }
            } else {
                // Video is taller - crop height
                let cropHeight = Int(videoSize.width / viewAspect)
                let cropGeometry = "\(Int(videoSize.width)):\(cropHeight)"
                cropGeometry.withCString { cString in
                    player.videoCropGeometry = UnsafeMutablePointer(mutating: cString)
                }
            }

        case .fill:
            // Stretch to fill - ignore aspect ratio
            let aspectRatio = "\(Int(bounds.width)):\(Int(bounds.height))"
            aspectRatio.withCString { cString in
                player.videoAspectRatio = UnsafeMutablePointer(mutating: cString)
            }
        }
    }

    var rate: Float = defaultPlayerRate {
        didSet {
            mediaPlayer?.rate = rate
        }
    }

    var time: Int = defaultPlayerTime

    var volume: Int = maxPlayerVolume {
        didSet {
            if options.hasAudioOption() {
                let error = ["error": "Audio disabled via options"]
                onEncounteredError(error)
            }

            let newVolume = max(minPlayerVolume, min(maxPlayerVolume, volume))
            oldVolume = newVolume

            if !mute {
                mediaPlayer?.audio?.volume = Int32(newVolume)
            }
        }
    }

    var mute: Bool = false {
        didSet {
            if options.hasAudioOption() {
                let error = ["error": "Audio disabled via options"]
                onEncounteredError(error)
            }

            let newVolume = mute ?
                minPlayerVolume :
                oldVolume

            mediaPlayer?.audio?.volume = Int32(newVolume)
            MediaPlayerManager.shared.setAppropriateAudioSession()
        }
    }

    var audioMixingMode: AudioMixingMode = .auto {
        didSet {
            MediaPlayerManager.shared.setAppropriateAudioSession()
        }
    }

    var shouldRepeat: Bool = false {
        didSet {
            if options.hasRepeatOption() {
                let error = ["error": "Repeat enabled via options"]
                onEncounteredError(error)
            }
        }
    }

    var playInBackground: Bool = false {
        didSet {
            MediaPlayerManager.shared.setAppropriateAudioSession()
        }
    }

    var autoplay: Bool = true {
        didSet {
            if !autoplay {
                options.append("--start-paused")
            }
        }
    }

    func play() {
        if let player = mediaPlayer {
            if options.hasStartPausedOption() {
                player.play()
            }

            player.play()
        }
    }

    func pause() {
        mediaPlayer?.pause()
    }

    func stop() {
        mediaPlayer?.stop()
    }

    func seek(_ value: Double, _ type: String) {
        if let player = mediaPlayer {
            if player.isSeekable {
                if type == "position" {
                    player.position = Float(value)
                } else {
                    player.time = VLCTime(int: Int32(value))
                }
            } else {
                if type == "position" {
                    time = Int(value * Double(mediaLength))
                } else {
                    time = Int(value)
                }
            }
        }
    }

    func postAction(_ action: Int) {
        if let dialog = vlcDialog, let reference = vlcDialogRef {
            dialog.postAction(Int32(action), forDialogReference: reference)
            vlcDialogRef = nil
        }
    }

    func dismiss() {
        if let dialog = vlcDialog, let reference = vlcDialogRef {
            dialog.dismissDialog(withReference: reference)
            vlcDialogRef = nil
        }
    }
}

private extension Array where Element == String {
    func hasAudioOption() -> Bool {
        let options = [
            "--no-audio",
            "-no-audio",
            ":no-audio",
        ]

        return contains { option in options.contains(option) }
    }
}

extension Array where Element == String {
    func hasRepeatOption() -> Bool {
        let options = [
            "--input-repeat=",
            "-input-repeat=",
            ":input-repeat=",
        ]

        return contains { option in
            options.contains { value in
                option.hasPrefix(value)
            }
        }
    }
}

extension Array where Element == String {
    func hasStartPausedOption() -> Bool {
        let options = [
            "--start-paused",
            "-start-paused",
            ":start-paused",
        ]

        return contains { option in options.contains(option) }
    }
}

extension Array where Element == String {
    mutating func removeStartPausedOption() {
        let options = [
            "--start-paused",
            "-start-paused",
            ":start-paused",
        ]

        removeAll { option in options.contains(option) }
    }
}

File: android/src/main/java/expo/modules/libvlcplayer/LibVlcPlayerView.kt
package expo.modules.libvlcplayer

import android.content.Context
import android.graphics.Matrix
import android.net.Uri
import android.view.TextureView
import android.view.ViewGroup
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import expo.modules.libvlcplayer.enums.AudioMixingMode
import expo.modules.libvlcplayer.enums.ContentFit
import expo.modules.libvlcplayer.records.Dialog
import expo.modules.libvlcplayer.records.MediaInfo
import expo.modules.libvlcplayer.records.MediaTracks
import expo.modules.libvlcplayer.records.Slave
import expo.modules.libvlcplayer.records.Track
import expo.modules.libvlcplayer.records.Tracks
import org.videolan.libvlc.LibVLC
import org.videolan.libvlc.Media
import org.videolan.libvlc.MediaPlayer
import org.videolan.libvlc.interfaces.IMedia
import org.videolan.libvlc.util.DisplayManager
import org.videolan.libvlc.util.VLCVideoLayout
import java.net.URI
import org.videolan.libvlc.Dialog as VLCDialog

const val DEFAULT_PLAYER_RATE: Float = 1f
const val DEFAULT_PLAYER_TIME: Int = 0
const val DEFAULT_PLAYER_SCALE: Float = 0f

const val MIN_PLAYER_VOLUME: Int = 0
const val MAX_PLAYER_VOLUME: Int = 100
const val PLAYER_VOLUME_STEP: Int = 10

private val DISPLAY_MANAGER: DisplayManager? = null
private val ENABLE_SUBTITLES: Boolean = true
private val USE_TEXTURE_VIEW: Boolean = true

class LibVlcPlayerView(
    context: Context,
    appContext: AppContext,
) : ExpoView(context, appContext) {
    private val playerView: VLCVideoLayout =
        VLCVideoLayout(context).also {
            addView(it)
        }

    internal var libVLC: LibVLC? = null
    internal var mediaPlayer: MediaPlayer? = null
    private var media: Media? = null
    internal var vlcDialog: VLCDialog? = null

    internal var mediaLength: Long = 0L
    internal var oldVolume: Int = MAX_PLAYER_VOLUME

    private var shouldCreate: Boolean = false
    internal var firstPlay: Boolean = false
    internal var firstTime: Boolean = false

    internal val onBuffering by EventDispatcher<Unit>()
    internal val onPlaying by EventDispatcher<Unit>()
    internal val onPaused by EventDispatcher<Unit>()
    internal val onStopped by EventDispatcher<Unit>()
    internal val onEndReached by EventDispatcher<Unit>()
    internal val onEncounteredError by EventDispatcher()
    internal val onTimeChanged by EventDispatcher()
    internal val onPositionChanged by EventDispatcher()
    internal val onESAdded by EventDispatcher<MediaTracks>()
    internal val onDialogDisplay by EventDispatcher<Dialog>()
    internal val onFirstPlay by EventDispatcher<MediaInfo>()
    internal val onBackground by EventDispatcher<Unit>()

    init {
        MediaPlayerManager.registerPlayerView(this)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        attachPlayer()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()

        detachPlayer()
    }

    fun createPlayer() {
        if (!shouldCreate) {
            return
        }

        destroyPlayer()

        val source = source ?: return

        if (autoplay) {
            options.removeStartPausedOption()
        }

        libVLC = LibVLC(context, options)
        setDialogCallbacks()
        mediaPlayer = MediaPlayer(libVLC)
        setMediaPlayerListener()

        try {
            URI(source)
        } catch (_: Exception) {
            val error = mapOf("error" to "Invalid source, media could not be set")
            onEncounteredError(error)
            return
        }

        media = Media(libVLC, Uri.parse(source))
        mediaPlayer!!.setMedia(media)
        media!!.release()

        addPlayerSlaves()

        mediaPlayer!!.play()

        shouldCreate = false
        firstPlay = true
        firstTime = true
    }

    fun attachPlayer() {
        mediaPlayer?.let { player ->
            val parent = playerView.getParent()

            if (parent == null) {
                addView(playerView)
            }

            val attached = player.getVLCVout().areViewsAttached()

            if (!attached) {
                player.attachViews(playerView, DISPLAY_MANAGER, ENABLE_SUBTITLES, USE_TEXTURE_VIEW)
            }
        }
    }

    fun detachPlayer() {
        mediaPlayer?.detachViews()
        removeAllViews()
    }

    fun destroyPlayer() {
        vlcDialog = null
        media = null
        mediaPlayer?.release()
        mediaPlayer = null
        libVLC?.release()
        libVLC = null
    }

    fun setPlayerTracks() {
        mediaPlayer?.let { player ->
            val audioTrack = tracks?.audio ?: player.getAudioTrack()
            val videoTrack = tracks?.video ?: player.getVideoTrack()
            val spuTrack = tracks?.subtitle ?: player.getSpuTrack()

            player.setAudioTrack(audioTrack)
            player.setVideoTrack(videoTrack)
            player.setSpuTrack(spuTrack)
        }
    }

    fun addPlayerSlaves() {
        slaves.forEach { slave ->
            val source = slave.source
            val type = slave.type
            val slaveType =
                if (type == "subtitle") {
                    IMedia.Slave.Type.Subtitle
                } else {
                    IMedia.Slave.Type.Audio
                }
            val selected = slave.selected ?: false

            try {
                URI(source)
            } catch (_: Exception) {
                val error = mapOf("error" to "Invalid slave, $type could not be added")
                onEncounteredError(error)
                return@forEach
            }

            mediaPlayer?.addSlave(slaveType, Uri.parse(source), selected)
        }
    }

    fun getMediaTracks(): MediaTracks {
        var mediaTracks = MediaTracks()

        mediaPlayer?.let { player ->
            val audioTracks = mutableListOf<Track>()
            val audios = player.getAudioTracks()

            audios?.forEach { track ->
                val trackObj = Track(id = track.id, name = track.name)
                audioTracks.add(trackObj)
            }

            val videoTracks = mutableListOf<Track>()
            val videos = player.getVideoTracks()

            videos?.forEach { track ->
                val trackObj = Track(id = track.id, name = track.name)
                videoTracks.add(trackObj)
            }

            val subtitleTracks = mutableListOf<Track>()
            val subtitles = player.getSpuTracks()

            subtitles?.forEach { track ->
                val trackObj = Track(id = track.id, name = track.name)
                subtitleTracks.add(trackObj)
            }

            mediaTracks =
                MediaTracks(
                    audio = audioTracks,
                    video = videoTracks,
                    subtitle = subtitleTracks,
                )
        }

        return mediaTracks
    }

    fun getMediaInfo(): MediaInfo {
        var mediaInfo = MediaInfo()

        mediaPlayer?.let { player ->
            val video = player.getCurrentVideoTrack()
            val pLength = player.getLength()
            val length =
                if (pLength != -1L) {
                    pLength
                } else {
                    0L
                }
            val seekable = player.isSeekable()
            val mediaTracks = getMediaTracks()

            mediaInfo =
                MediaInfo(
                    width = video?.width ?: 0,
                    height = video?.height ?: 0,
                    length = length.toDouble(),
                    seekable = seekable,
                    tracks = mediaTracks,
                )

            mediaLength = length
        }

        return mediaInfo
    }

    fun setupPlayer() {
        mediaPlayer?.let { player ->
            setPlayerTracks()

            if (volume != MAX_PLAYER_VOLUME || mute) {
                val newVolume =
                    if (mute) {
                        MIN_PLAYER_VOLUME
                    } else {
                        volume
                    }

                player.setVolume(newVolume)
            }

            if (rate != DEFAULT_PLAYER_RATE) {
                player.setRate(rate)
            }

            if (time != DEFAULT_PLAYER_TIME) {
                player.setTime(time.toLong())
            }

            if (scale != DEFAULT_PLAYER_SCALE) {
                player.setScale(scale)
            }

            if (aspectRatio != null) {
                player.setAspectRatio(aspectRatio)
            }

            applyContentFit()

            time = DEFAULT_PLAYER_TIME
        }
    }

    var source: String? = null
        set(value) {
            val old = field
            field = value
            shouldCreate = value != old
        }

    var options: ArrayList<String> = ArrayList()
        set(value) {
            val old = field
            field = value
            shouldCreate = value != old
        }

    var tracks: Tracks? = null
        set(value) {
            field = value
            setPlayerTracks()
        }

    var slaves: ArrayList<Slave> = ArrayList()
        set(value) {
            val newSlaves = value.filter { slave -> slave !in field }

            field = field.apply { addAll(newSlaves) }

            if (!newSlaves.isEmpty()) {
                addPlayerSlaves()
            }
        }

    var scale: Float = DEFAULT_PLAYER_SCALE
        set(value) {
            field = value
            mediaPlayer?.setScale(value)
        }

    var aspectRatio: String? = null
        set(value) {
            field = value
            mediaPlayer?.setAspectRatio(value)
        }

    var contentFit: ContentFit = ContentFit.CONTAIN
        set(value) {
            field = value
            applyContentFit()
        }

    fun applyContentFit() {
        val viewWidth = playerView.width.toFloat()
        val viewHeight = playerView.height.toFloat()

        if (viewWidth <= 0 || viewHeight <= 0) return

        val textureView = findTextureView(playerView) ?: return

        when (contentFit) {
            ContentFit.CONTAIN -> {
                textureView.setTransform(Matrix())
            }

            ContentFit.COVER -> {
                val player = mediaPlayer ?: return
                val videoTrack = player.getCurrentVideoTrack() ?: return

                val videoWidth = videoTrack.width.toFloat()
                val videoHeight = videoTrack.height.toFloat()

                if (videoWidth <= 0 || videoHeight <= 0) return

                val videoAspect = videoWidth / videoHeight
                val viewAspect = viewWidth / viewHeight

                val matrix = Matrix()

                val scale: Float =
                    if (videoAspect > viewAspect) {
                        videoAspect / viewAspect
                    } else {
                        viewAspect / videoAspect
                    }

                matrix.setScale(scale, scale, viewWidth / 2f, viewHeight / 2f)
                textureView.setTransform(matrix)
            }

            ContentFit.FILL -> {
                val player = mediaPlayer ?: return
                val videoTrack = player.getCurrentVideoTrack() ?: return

                val videoWidth = videoTrack.width.toFloat()
                val videoHeight = videoTrack.height.toFloat()

                if (videoWidth <= 0 || videoHeight <= 0) return

                val videoAspect = videoWidth / videoHeight
                val viewAspect = viewWidth / viewHeight

                val matrix = Matrix()

                val scaleX: Float
                val scaleY: Float

                if (videoAspect > viewAspect) {
                    scaleX = 1f
                    scaleY = videoAspect / viewAspect
                } else {
                    scaleX = viewAspect / videoAspect
                    scaleY = 1f
                }

                matrix.setScale(scaleX, scaleY, viewWidth / 2f, viewHeight / 2f)
                textureView.setTransform(matrix)
            }
        }
    }

    private fun findTextureView(parent: ViewGroup): TextureView? {
        for (i in 0 until parent.childCount) {
            val child = parent.getChildAt(i)
            if (child is TextureView) return child
            if (child is ViewGroup) {
                val found = findTextureView(child)
                if (found != null) return found
            }
        }
        return null
    }

    var rate: Float = DEFAULT_PLAYER_RATE
        set(value) {
            field = value
            mediaPlayer?.setRate(value)
        }

    var time: Int = DEFAULT_PLAYER_TIME

    var volume: Int = MAX_PLAYER_VOLUME
        set(value) {
            field = value

            if (options.hasAudioOption()) {
                val error = mapOf("error" to "Audio disabled via options")
                onEncounteredError(error)
            }

            val newVolume = value.coerceIn(MIN_PLAYER_VOLUME, MAX_PLAYER_VOLUME)
            oldVolume = newVolume

            if (!mute) {
                mediaPlayer?.setVolume(newVolume)
            }
        }

    var mute: Boolean = false
        set(value) {
            field = value

            if (options.hasAudioOption()) {
                val error = mapOf("error" to "Audio disabled via options")
                onEncounteredError(error)
            }

            val newVolume =
                if (value) {
                    MIN_PLAYER_VOLUME
                } else {
                    oldVolume
                }

            mediaPlayer?.setVolume(newVolume)
            MediaPlayerManager.audioFocusManager.updateAudioFocus()
        }

    var audioMixingMode: AudioMixingMode = AudioMixingMode.AUTO
        set(value) {
            field = value
            MediaPlayerManager.audioFocusManager.updateAudioFocus()
        }

    var repeat: Boolean = false
        set(value) {
            field = value

            if (options.hasRepeatOption()) {
                val error = mapOf("error" to "Repeat enabled via options")
                onEncounteredError(error)
            }
        }

    var playInBackground: Boolean = false
        set(value) {
            field = value
            MediaPlayerManager.audioFocusManager.updateAudioFocus()
        }

    var autoplay: Boolean = true
        set(value) {
            field = value

            if (!value) {
                options.add("--start-paused")
            }
        }

    fun play() {
        mediaPlayer?.let { player ->
            if (options.hasStartPausedOption()) {
                player.play()
            }

            player.play()
        }
    }

    fun pause() {
        mediaPlayer?.pause()
    }

    fun stop() {
        mediaPlayer?.stop()
    }

    fun seek(
        value: Double,
        type: String,
    ) {
        mediaPlayer?.let { player ->
            if (player.isSeekable()) {
                if (type == "position") {
                    player.setPosition(value.toFloat())
                } else {
                    player.setTime(value.toLong())
                }
            } else {
                if (type == "position") {
                    time = (value * mediaLength.toDouble()).toInt()
                } else {
                    time = value.toInt()
                }
            }
        }
    }

    fun postAction(action: Int) {
        vlcDialog?.let { dialog ->
            when (dialog) {
                is VLCDialog.QuestionDialog -> {
                    dialog.postAction(action)
                    vlcDialog = null
                }
            }
        }
    }

    fun dismiss() {
        vlcDialog?.let { dialog ->
            dialog.dismiss()
            vlcDialog = null
        }
    }

    private fun ArrayList<String>.hasAudioOption(): Boolean {
        val options =
            setOf(
                "--no-audio",
                "-no-audio",
                ":no-audio",
            )

        return this.any { option -> option in options }
    }

    internal fun ArrayList<String>.hasRepeatOption(): Boolean {
        val options =
            setOf(
                "--input-repeat=",
                "-input-repeat=",
                ":input-repeat=",
            )

        return this.any { option ->
            options.any { value ->
                option.startsWith(value)
            }
        }
    }

    internal fun ArrayList<String>.hasStartPausedOption(): Boolean {
        val options =
            setOf(
                "--start-paused",
                "-start-paused",
                ":start-paused",
            )

        return this.any { option -> option in options }
    }

    internal fun ArrayList<String>.removeStartPausedOption() {
        val options =
            setOf(
                "--start-paused",
                "-start-paused",
                ":start-paused",
            )

        this.removeAll(options)
    }
}
